### PR TITLE
Work branch

### PR DESCRIPTION
**Pull request by eliasnaur**
Work branch, containing

- fix for non-uniform barrier() invocation when shaders run out of memory
- use of mediump precison in k4 to boost performance on older Androids.

### CODE REVIEW

**Review by raphlinus — COMMENTED** [review]
Ok, I'm in favor of this change, and also the mediump change. I was hoping to get time to review the control flow more carefully, but haven't done so yet. Likely I'll get a chance to take another look next time I go into those shaders.

On the mediump issue I still don't have clarity on whether it can be expected to have any effect in Vulkan-land. Relevant resources include:

* https://github.com/KhronosGroup/Vulkan-Docs/issues/706
* https://stackoverflow.com/questions/49887574/how-to-get-16bit-floats-in-modern-glsl-with-vulkan
* https://blogs.igalia.com/itoral/2018/12/04/vk_khr_shader_float16_int8-on-anvil/
* https://www.lunarg.com/wp-content/uploads/2019/09/Automatic-RelaxedPrecision-Decoration-and-Conversion-in-Spirv-Opt_r1.pdf

Of these, the last is perhaps most interesting. It suggests an explicit spirv-opt pass to convert mediump annotations into explicit f16 operations (as enabled through the VK_KHR_shader_float16_int8 extension). That looks like something that's worth experimenting with.

Certainly on Pixel 4, I see no performance difference at all applying the patch, or (previously) adding mediump to "hot" variables. I'm not sure why yet. In any case, worst case the patch is probably harmless, just acting as a no-op when the compiler ignores the RelaxedPrecision decorations.

**Review by raphlinus — APPROVED** [review]

File: piet-gpu/shader/elements.comp [hunk 1]
```diff
@@ -392,7 +388,6 @@ void main() {
                 vec2 lw = get_linewidth(st);
                 anno_begin_clip.linewidth = st.linewidth * sqrt(abs(st.mat.x * st.mat.w - st.mat.y * st.mat.z));
             } else {
-                anno_begin_clip.bbox = begin_clip.bbox;
```
[the remark this answers is not in the document]
Comment (eliasnaur) [reply piet-gpu/shader/elements.comp:395]: Intentional, but unrelated. I extracted its removal into a separate https://github.com/linebender/piet-gpu/commit/3b4a72deb9abc5665cb725bb43fa330b062f4581.

**Comment by eliasnaur** [comment]
Merged.

**Comment by eliasnaur** [comment]
> On the mediump issue I still don't have clarity on whether it can be expected to have any effect in Vulkan-land. Relevant resources include:
> 
>     * [KhronosGroup/Vulkan-Docs#706](https://github.com/KhronosGroup/Vulkan-Docs/issues/706)
> 
>     * https://stackoverflow.com/questions/49887574/how-to-get-16bit-floats-in-modern-glsl-with-vulkan
> 
>     * https://blogs.igalia.com/itoral/2018/12/04/vk_khr_shader_float16_int8-on-anvil/
> 
>     * https://www.lunarg.com/wp-content/uploads/2019/09/Automatic-RelaxedPrecision-Decoration-and-Conversion-in-Spirv-Opt_r1.pdf
> 
> 
> Of these, the last is perhaps most interesting. It suggests an explicit spirv-opt pass to convert mediump annotations into explicit f16 operations (as enabled through the VK_KHR_shader_float16_int8 extension). That looks like something that's worth experimenting with.
> 

There are two mechanisms at play:

- Precision qualifiers, mapped to [RelaxedPrecision](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_relaxedprecisionsection_a_relaxed_precision) in (core) SPIR-V. They act like hints in that the types are still 32-bit floats, but the driver is free to do computation with fewer bits of precision (at least 16 bits according to spec).
- Explicit 16-bit float types, enabled by extensions such as VK_KHR_shader_float16_int8.

Your [last reference](https://www.lunarg.com/wp-content/uploads/2019/09/Automatic-RelaxedPrecision-Decoration-and-Conversion-in-Spirv-Opt_r1.pdf) discusses an optimization pass to convert precision qualifiers to 16-bit floats. This only makes sense because some drivers such as MoltenVK ignores precision qualifiers, yet supports 16-bit floats:

>Unfortunately, some Vulkan drivers such as MoltenVK ignore the RelaxedPrecision directive, usually because the underlying implementation (in the case of MoltenVK, Metal), does not support the RelaxedPrecision semantics. If it is the situation, however, that the underlying implementation does support a true 16-bit floating point type (as it is with Metal), benefits of RelaxedPrecision semantics can be achieved by converting RelaxedPrecision instructions to true 16-bit-type instructions using the ​VK_KHR_shader_float16_int8​ extension.

(It seems to me MoltenVK should automatically apply this optimization, but I digress.)

Now, given that my Pixel 1 does seem to use RelaxedPrecision it would surprise me if your Pixel 4 ignores precision qualifiers while supporting 16-bit floats. But I can't say for sure, of course.

**Comment by raphlinus** [comment]
I ran some experiments on Pixel 4 using this commandline:

```
spirv-opt --convert-relaxed-to-half kernel4.spv -o kernel4.spv
```

The results were somewhat inconclusive but were consistent with the driver respecting RelaxedPrecision but it not making a significant performance difference on that hardware. That is surprising but it at least gives us more knobs to play with.

**Pull request closed without merging**